src/wp-includes/pluggable.php

```diff
@@ -2624,90 +2624,186 @@ function wp_hash( $data, $scheme = 'auth', $algo = 'md5' ) {
 	 * instead use the other package password hashing algorithm.
 	 *
 	 * @since 2.5.0
+	 * @since x.y.z The password is now hashed using bcrypt instead of phpass.
 	 *
-	 * @global PasswordHash $wp_hasher PHPass object.
+	 * @global PasswordHash $wp_hasher phpass object.
 	 *
 	 * @param string $password Plain text user password to hash.
 	 * @return string The hash string of the password.
 	 */
 	function wp_hash_password( $password ) {
 		global $wp_hasher;
 
-		if ( empty( $wp_hasher ) ) {
-			require_once ABSPATH . WPINC . '/class-phpass.php';
-			// By default, use the portable hash from phpass.
-			$wp_hasher = new PasswordHash( 8, true );
+		if ( ! empty( $wp_hasher ) ) {
```

Comment:
@johnbillion 

I feel like this has the potential for a lot of breakage.

IIRC, a **ton** of plugins that do custom user creation and/or profile updates do checks like "if empty, create new password hash instance".

This opens up the possibility that password hashes are created with $wp_hasher, but not validated with it, since those same plugins would not re-instantiate on the fly during validation.

Reply:
See:

https://wpdirectory.net/search/01JH120WAY9FXBHJFMT7QN5HR5


![image](https://github.com/user-attachments/assets/40879d6a-cab0-445b-8cdb-c1bd74559ee0)

Reply:
Ok, so we can boil that down to: plugins that manually hash a password with `PasswordHash` and then let core do the hash validation at a later date. That should be supported by this PR and not be a problem because `wp_check_password()` will check a password with `PasswordHash` when its hash starts with `$P$`, regardless of the `$wp_hasher` global.

The problem you describe would only take effect if a plugin specifically assigns a hasher instance to the `$wp_hasher` global (the instance isn't a singleton).

Am I understanding correctly?

Search results for instances of that here: https://wpdirectory.net/search/01JH19AV75Q6TJA6VRZ0A9EMG7

Reply:
The reverse case is also true, there are plugins calling $wp_hash->checkPassword()

It's not immediately clear to me what the implications are. There are several scenarios that'd need to be taken into account.

Reply:
I have this documented in the Fortress docs [here](https://github.com/snicco/fortress/blob/beta/docs/modules/password/password-hashing.md#doing-it-wrong-01)

There's nothing that can be done to support this particular edge case.

For me, that was an acceptable trade-off. But the scale is obviously different.

Reply:
I'll do some digging through plugins which are using `PasswordHash` directly and/or messing with the `$wp_hasher` global.

Reply:
Are there any follow-up actions needed, or is this resolved now?